src/config.ts

```diff
@@ -0,0 +1,30 @@
+import * as Dockerode from 'dockerode';
+
+interface AppConfig {
```

Comment:
do you know the difference between `type` and `interface`

Reply:
no not really except some murky stuff about merging definitions on interface. I could probably just as easily make these type aliases.

for funnies, in other languages there _is_ significant difference. for example, Haskell has `type`, `data`, and `newtype` which are pretty cool if a bit complex

`type` creates a new type as a simple or compound composition of other types.

```haskell
type Vector = ( Int, Float, [Char] ) {- magnitude, direction, name -}
type String = [Char]
type Amount = Int
```

these are like structural types and "if it fits" you can pass them around.
for instance, if I have a three-tuple of `int`, `float`, and `[char]` then I
can pass it as if it were a `Vector` because it matches the same structure.

`data` creates a data constructor that creates tagged-types. these types
can contain other data types in them. without other data it's a sum or union
data type, but if we add additional data then at least one of the value becomes
a product type.

```haskell
data Direction = Forward | Reverse {- just a simple sum type -}
data Currency
  = USD Int Int {- dollars, cents -}
  | JPY Int
```

These types are _stricter_ in that they _must_ carry their tag or
they won't work. You can't pass in `5` as a `JPY`. It _must_
be constructed as `JPY 5` or in dollars `USD 5 0`.

But the `data` types don't share the same kind of type erasure as `type`
does. It has to track the type at runtime and so these are _boxed_ types.
They add overhead.

`newtype` is the hybrid. It forces the use of the tag but it's limited to a
single data type constructor, to not being a sum type. As a result it erases
completely at runtime and so can be used to force constraints without
overhead.

```haskell
newtype Email = Email String
```

In this example we see it very similar to how `type` was declared. The
difference is that we carry a tag and so we can't just pass any string
where an email is required.

```haskell
type Email = String
newtype Email' = Email String

isEmail :: Email -> Bool
isEmail e = e /= ""

isEmail' :: Email' -> Bool
isEmail' (Email e) -> e /= ""

isEmail "test@example.com" {- True -}
isEmail' "test@example.com" {- compile error -}
isEmail' (Email "test@example.com") {- True -}
```

with `newtype` we can force runtime constraints such
as requiring that a certain value be created through an
internal function which enforces value rules.

```haskell
module EmailUtil where (ValidEmail, makeEmail)

newtype ValidEmail = Email String

makeEmail :: String -> Maybe ValidEmail
makeEmail s = if s /= "" then Just (Email s) else Nothing
```

Note that in this example I have shown that the _type constructor name_
on the right of the type doesn't have to match the name of
the type on the left side. This module exports the type but
not the constructor; that means that there's no way to create
the type from the outside unless you use `makeEmail` which
will return `Nothing` if it doesn't validate the input string - this
can be incredibly useful for preventing the propagation of invalid
values deep inside a system because _if_ we get a type `ValidEmail`
then we are guaranteed that our system considers it valid and that
it did not come directly from external input.